### PR TITLE
ingestion udf schema change sample

### PR DESCRIPTION
**Pull request by chenqin**
In short term, we want to be able take 
- nested row that not yet supported by nebula and convert to flat rows
- timestamp and other datatypes to unix_timestamp
- string and binary to numerical value if possible to save space
- downsample logic based on percentage of a column and filter by event type 

Limitations:
- don't support udf taking two or more columns
- DataType system still seems not incorporate with ansi sql while rest of structured data systems are.

@shawncao what do you think

### CODE REVIEW

**Comment by shawncao** [comment]
Hey, wondering why do you close this? Very interesting topic to chat about, I thought we should take some time to think and discuss it.

**Comment by chenqin** [comment]
> Hey, wondering why do you close this? Very interesting topic to chat about, I thought we should take some time to think and discuss it.

It was human error

**Comment by shawncao** [comment]
This "feature" has lots of different variants in implementation, I can think of
- Transformer: column level spec of transformation in ingestion time. (such as Druid ingestion spec)
- Funnel: maintain multiple data set for the same data source, one is usually a rollup on top of the other.
- View: materialized query cache for query time speed.
- Sampling: downsampling logic in either ingestion time or query time or both.

I think these variants are basically addressing many scenarios with overlap - meaning they may be able to solve the same problems in real life.

I think it's interesting because of these two aspects: 
- how do we think what is the most powerful way to address the issue to meet majority of users' need?
- how do we know which is the simplest and easiest approach in terms of user's experience?

Also, we should think about which is best for Nebula in differentiating itself from others, eventually we may implemented a hybrid solution wrapping the good parts of all them. 

Another thing we can take advantage is Nebula SDK, how Nebula itself can leverage the capability of the JS engine to transform existing data into any other types, should we allow user write JS code in the transformer spec directly?

Just some rough thoughts as a start, let's follow up on this topic and we should be able to narrow down to the thing we want to build eventually.

**Comment by shawncao** [comment]
@chenqin I kinda found the PR content is for introducing retention unit in our spec. To NOT lose the big topic of discussions on data funnel/transformation/sampling/view/filtering, could you separate them into different threads?

Keep this PR and copy our discussion into an Issue or Discussion? (I found Discussions is new in github repo, haven't tried it yet officially)

**Comment by chenqin** [comment]
>     * l: maintain multiple data set for the same data source, one is usually a rollup on top of the other.

My assumption is 
- Nebula would be most effective slice/dice on single table per query on top of immutable rows. 
- Data processing before landing data to Nebula best split from query engine based on goal difference.

In short term, my plan is to add simple UDF and support sync data from source with nested or complex schema nebula haven't support. (flatten schema can help query optimizer figure out column pruning etc) As I demoed before, user can utilize streaming processing or batch processing engine to per-process datasets. Here is the catch, they need experts and step learning curve and operation curve even we provide easy integration story.

In mid term, I see OLAP as part of data system stack evolved quite a bit in last couple of years. In most recent, folks recognize there always need some online compute capability to shape input datasets without spin off a micro service. such as ksql https://www.confluent.io/blog/ksql-streaming-sql-for-apache-kafka/ or hologres https://www.hologres.io/ Ingestion UDF is small step follow this trend into future.

**Comment by shawncao** [comment]
your short term sounds good to me, my point is this area is huge, we can play very well in this area if we do a well-rounded thinking and design.

File: src/configs/test.yml [hunk 1]
```diff
@@ -27,6 +27,15 @@ tables:
     max-mb: 10000
     max-hr: 240
     schema: "ROW<id:int, event:string, items:list<string>, flag:bool, value:tinyint>"
+    transformer:
+      schema: "ROW<id:int, event:string, item:string, flag:bool, value:tinyint, ts:long>"
```
Comment (shawncao) [inline src/configs/test.yml:31]: Just to keep on the same page, Nebula supports compound types https://github.com/varchar-io/nebula/blob/master/src/type/Type.h#L43-L45, it's just we have bugs to fix to make them work end-to-end.

**Pull request closed without merging**